connect-inject/health_check_resource.go

```diff
@@ -0,0 +1,283 @@
+package connectinject
+
+import (
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+	"golang.org/x/net/context"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	// labelInject is the label which is applied by the connect-inject webhook to all pods.
+	// This is the key the controller will use on the label filter for its lister, watcher and reconciler.
+	labelInject = "consul.hashicorp.com/connect-inject-status"
+
+	// kubernetesSuccessReasonMsg will be passed for passing health check's Reason to Consul
+	kubernetesSuccessReasonMsg = "Kubernetes Health Checks Passing"
+)
+
+type HealthCheckResource struct {
+	Log                 hclog.Logger
+	KubernetesClientset kubernetes.Interface
+
+	// ConsulUrl holds the url information for client connections
+	ConsulUrl *url.URL
+	// ReconcilePeriod is the period by which reconcile gets called, default to 1 minute
+	ReconcilePeriod time.Duration
+
+	Ctx  context.Context
+	lock sync.Mutex
+}
+
+// Run is the long-running runloop for periodically running Reconcile
+// it initially starts a Reconcile phase at startup and then calls Reconcile
+// once every ReconcilePeriod time
+func (h *HealthCheckResource) Run(stopCh <-chan struct{}) {
+	// Start the background watchers
+	err := h.Reconcile()
+	if err != nil {
+		h.Log.Error("reconcile returned an error", "err", err)
+	}
+
+	reconcileTimer := time.NewTimer(h.ReconcilePeriod)
+	defer reconcileTimer.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			h.Log.Info("HealthCheckController quitting")
+			return
+
+		case <-reconcileTimer.C:
+			err = h.Reconcile()
+			if err != nil {
+				h.Log.Error("reconcile returned an error", "err", err)
+			}
+			reconcileTimer.Reset(h.ReconcilePeriod)
+		}
+	}
```

Comment:
could we add tests for the behavior in this function?

Reply:
Done!